### ISSUE TITLE
decide whether can_dig should be called with a nil player or not

### ISSUE THREAD

**Issue by t4im** [minetest/minetest#2108]
Currently some mods like sign_lib expect the player that is passed to `can_dig` never to be nil.
This raised among other things minetest-technic/technic#172 

However this has been official but also undocumented behavior so far, due to minetests `dig_node` calling `can_dig` allready with a nil player (also currently from master minetest_game, although minetest/minetest_game#391 is about to change that)

Up to you now to decide whether `dig_node` needs an api overhaul or the behavior requires documentation. :)

Keep in mind, that `dig_node` currently does not uphold its complete documented behavior anyway: minetest/minetest#2015

**Comment by 4aiman** [comment]
*TL;DR:*
IMHO, it's up to RBA to add a simple check (as far as the "technic" mod is concerned). In fact, right now he's able to decide whether mobs can dig on his own - and that's great!

*The long explanation*
There won't be a player for you if a MOB tries to dig some stuff and the game checks whether it should.

One can call minetest.remove_node(pos) to avoid any checks.
But some mods (protection, some mobs/bots ones) have certain rules for what an entity (or a node like TNT upon explosion) can dig and what it can't.

Say, I have some node that should dig blocks around in the way a player does it (e.g. with all appropriate effects - multiplication, protection violation de-buffs etc). Someone may set up such a node to try to destroy someone else's building and logout to remain uncaught. 
So, we can't rely on the minetest.get_player_by_name() - there won't be such a player. 
However, we can "fake" the player variable, so all the other mods will work with that structure as if it were the real player him/her~self (e.g. ban him/her, punish, schedule some drawbacks to be applied to a cunning mischievous griever etc).

Of course, the above mod should be very complex and deal with every copied objectref player-only method, but that's not a reason to devoid modders of such a great way to make player-like mobs/nodes.

**Comment by yo-we** [comment]
4aiman: Just for better comprehension for me: Does that imply that the technics machine "Quarry" will be a player in terms of code?

**Comment by t4im** [comment]
It already kind of is. But that is not the issue at hand. Even with every single mod compensating for this issue (which they already have to do or crash trying not to, as some others), theres still the question on whether or not nil player passing should be allowed or not. This is an api decision.

One option leading to updated documentation and mods consequently having to add nil checks if they have not already, another leading to changing api around `dig_node` stopping to pass nil players, thereby stoping to break it.

Ignoring this issue as "not our problem" or "this is compensatable" is _not_ a solution. It _is_ a minetest api bug, that should be fixed eventually, as it caused and will continuously continue to cause preventable crashs. Sure each single one can be attributed first to the respective mod. But is in the end an api fault.
Yet apparently it was overseen so far.
@ShadowNinja mind labeling it at least, to make it easier findable? ;) maybe that will raise awareness.

**Comment by ShadowNinja** [comment]
I vote for making `nil` a valid value, since the alternative is having every mod that does this construct a fake player.

**Comment by nerzhul** [comment]
agree too, this is a good point. It's not useful to allocate a fake player

**Comment by 4aiman** [comment]
*TL;DR*
+1 for <code>nil</code> being a valid value.

*The long one*
@yo-we : that was just an example. 
I don't want to copy all fields of the objectref, but some of those - a mere essentials to make the code more uniform. NO actual allocation of a CAO/SAO was implied!

Also, I simply cannot agree with t4im that this should be "fixed".

Those, who want, should add a digger.
Those, who don't care - should be able to ignore that parameter.
Even in a game with "protection and stuff" features there might be entities/events/nodes that would destroy blocks regardless of restriction. 

Why should an "explody" mob care for the name of the owner of the stuff it is going to destroy?

I've mentioned "remove node" func, true that it's possible to use that.
But why should anyone bother to get an "old_node"?
How's that better than adding a check for nil digger?

**Comment by t4im** [comment]
@4aiman i think you misread me; or can you really not agree that we should prevent crashes caused by unclear api contracts? ;-)

**Comment by ShadowNinja** [comment]
e680d49774b0fa2d71ac5de12fb116fe9b886132
